python-checks/src/main/java/org/sonar/python/checks/StringFormatMisuseCheck.java

```diff
@@ -41,32 +43,50 @@
   public void initialize(Context context) {
     context.registerSyntaxNodeConsumer(Tree.Kind.MODULO, ctx -> {
       BinaryExpression expression = (BinaryExpression) ctx.syntaxNode();
-      if (!expression.leftOperand().is(Tree.Kind.STRING_LITERAL)) {
+      StringLiteral formatString = extractStringLiteral(expression.leftOperand());
+      if (formatString == null) {
         return;
       }
 
-      StringLiteral formatString = ((StringLiteral) expression.leftOperand());
-      Optional<StringFormat> formatOptional = StringFormat.createFromPrintfStyle(ctx, formatString, formatString.trimmedQuotesValue());
-
+      Optional<StringFormat> formatOptional = StringFormat.createFromPrintfStyle(ctx, expression.leftOperand(), formatString, formatString.trimmedQuotesValue());
       if (!formatOptional.isPresent()) {
         // The string format contains invalid syntax.
         return;
       }
 
       StringFormat format = formatOptional.get();
-      if (expression.rightOperand().is(Tree.Kind.TUPLE)) {
-        checkTuples(ctx, format, ((Tuple) expression.rightOperand()));
-      } else if (expression.rightOperand().is(Tree.Kind.DICTIONARY_LITERAL)) {
-        checkDictionaries(ctx, format, ((DictionaryLiteral) expression.rightOperand()));
-      } else if (format.numExpectedArguments() == 1) {
-        format.replacementFields().get(0).validateArgument(expression.rightOperand());
+
+      Expression rhs = expression.rightOperand();
+      if (format.hasNamedFields()) {
+        checkNamed(ctx, format, rhs);
+      } else {
+        checkPositional(ctx, format, rhs);
```

Comment:
I'm not sure if the fix should be here but the current implementation raises an FP for:
```
"%%(key)s" % {"other": "str"}  # OK. Pylint FP
```
The problem is that `createFromPrintfStyle` returns a `StringFormat` with 0 replacement fields, so we enter `checkPositional` and then raise because the rhs is not a tuple.

Either the `StringFormat` should return `Optional.empty()` in such case, or this logic should be updated.